### PR TITLE
Added boundary feature to ReorderableList.

### LINKED ISSUE #146112: Proposal to set boundary for the `ReorderableListView` when dragging item
### Use case

ReorderableListView don't has API for limit drag area.
Simulator Screen Recording - iPhone 15 Plus - 2024-04-02 at 13 13 21

### Proposal

ReorderableListView should provide api to limit drag area.

### PR DESCRIPTION
**Pull request by yiiim**
Fixes: #146112

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
    home: Scaffold(
      body: Center(
        child: Container(
          width: 200,
          color: Colors.green,
          child: DragBoundary(
            child: CustomScrollView(
              shrinkWrap: true,
              slivers: <Widget>[
                SliverReorderableList(
                  itemBuilder: (BuildContext context, int index) {
                    return ReorderableDragStartListener(
                      key: ValueKey<int>(index),
                      index: index,
                      child: Text('$index'),
                    );
                  },
                  itemCount: 5,
                  onReorder: (int fromIndex, int toIndex) {},
                ),
              ],
            ),
          ),
        ),
      ),
    ),
  ));
}

```

</details>

| without `DragBoundary` |  with the `DragBoundary` | 
| ------- | ------- 
| Screen Recording 2024-04-03 at 13 01 19 | Screen Recording 2024-04-03 at 13 01 19 |

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:1584]: This feels a bit difficult for users to define. Say the user wants to stick within the bounds of the Viewport, since a Viewport will expand to fill all of the available space, how does the user know what that rect is?

**Review by Piinks — COMMENTED** [review]
Hi @yiiim! Thanks for your patience, and thanks for the PR!

**Comment by yiiim** [comment]
I implemented a generic drag gesture handling here: #147521. This PR needs to be reimplemented after it.

**Comment by goderbauer** [comment]
(triage): We are still waiting on https://github.com/flutter/flutter/pull/147521 before we can make progress here.

**Comment by chunhtai** [comment]
(triage) converting to draft for now until the other pr is merged. Feel free to convert back once this is ready for review

**Comment by chunhtai** [comment]
(triage) still waiting for the other pr to merge

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:102]: What do you think if we make this the default behavior, so that it's opt-out instead of opt-in?

Comment (yiiim) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:102]: I have considered doing this. If the boundary widget is named `ReorderableDragBoundary`, I think it is feasible. `DragBoundary` is not suitable to be provided by default to `SliverReorderableList`, as they do not seem to have any direct connection.

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:1392]: Have you considered having `ReorderableList` inherit the drag boundary through `forRectMaybeOf` by default instead of passing it down as a parameter? Or maybe it should do both, inherit by default, and have a parameter that overrides the default inherited provider.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:1392]: I'm also curious if there is any native behavior we have to compare to on other platforms (if a similar component exists).

Comment (yiiim) [reply packages/flutter/lib/src/widgets/reorderable_list.dart]: Modified to default to using `DragBoundary.forRectMaybeOf`.

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Review by Renzo-Olivares — APPROVED** [review]
LGTM, thank you for the contribution!

**Pull request merged**
